src/go/serviceconfig/service_config_fetcher.go

```diff
@@ -0,0 +1,252 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package serviceconfig
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"io/ioutil"
+	"math"
+	"net/http"
+	"time"
+
+	"github.com/GoogleCloudPlatform/esp-v2/src/go/metadata"
+	"github.com/GoogleCloudPlatform/esp-v2/src/go/options"
+	"github.com/GoogleCloudPlatform/esp-v2/src/go/util"
+	"github.com/golang/glog"
+	"github.com/golang/protobuf/proto"
+
+	confpb "google.golang.org/genproto/googleapis/api/serviceconfig"
+	smpb "google.golang.org/genproto/googleapis/api/servicemanagement/v1"
+)
+
+type ServiceConfigFetcher struct {
+	serviceName         string
+	checkRolloutsTicker *time.Ticker
+	client              http.Client
+	mf                  *metadata.MetadataFetcher
+	opts                options.ConfigGeneratorOptions
+
+	curServiceConfig    *confpb.Service
+	curRolloutId        string
+	curFetchConfigError error
+}
+
+func NewServiceConfigFetcher(mf *metadata.MetadataFetcher, opts options.ConfigGeneratorOptions, serviceName, configId string) (*ServiceConfigFetcher, error) {
+	caCert, err := ioutil.ReadFile(opts.RootCertsPath)
+	if err != nil {
+		return nil, err
+	}
+	caCertPool := x509.NewCertPool()
+	caCertPool.AppendCertsFromPEM(caCert)
+	scf := &ServiceConfigFetcher{
+		client: http.Client{
+			Transport: &http.Transport{
+				TLSClientConfig: &tls.Config{
+					RootCAs: caCertPool,
+				},
+			},
+			Timeout: opts.HttpRequestTimeout,
+		},
+		serviceName: serviceName,
+		mf:          mf,
+		opts:        opts,
+	}
+	_, _ = scf.fetchConfig(configId)
+	return scf, nil
+}
+
+func (scf *ServiceConfigFetcher) CurServiceConfig() (*confpb.Service, error) {
+	return scf.curServiceConfig, scf.curFetchConfigError
+}
+
+func (scf *ServiceConfigFetcher) SetFetchConfigTimer(interval *time.Duration, callback func(serviceConfig *confpb.Service)) {
+	go func() {
+		glog.Infof("start checking new rollouts every %v seconds", *interval)
+		scf.checkRolloutsTicker = time.NewTicker(*interval)
+
+		for range scf.checkRolloutsTicker.C {
+			glog.Infof("check new rollouts for service %v", scf.serviceName)
+
+			serviceConfig, err := scf.fetchConfig("")
+			if err != nil {
+				glog.Errorf("error occurred when checking new rollouts, %v", err)
+				continue
+
+			}
+
+			if serviceConfig != nil {
+				callback(serviceConfig)
+			}
+		}
+	}()
+}
+
+func (scf *ServiceConfigFetcher) CurRolloutId() string {
+	return scf.curRolloutId
+}
+
+func (scf *ServiceConfigFetcher) curConfigId() string {
+	if scf.curServiceConfig == nil {
+		return ""
+	}
+	return scf.curServiceConfig.Id
+}
+
+// Fetch the service config by given configId. If configId is empty, try to
+// fetch the latest service config,.
+func (scf *ServiceConfigFetcher) fetchConfig(configId string) (*confpb.Service, error) {
+	_fetchConfig := func(configId string) (*confpb.Service, error) {
```

Comment:
What is the point of this internal function? It's just called like a normal function right after, so I don't see why it's needed.

Reply:
It is implementation of fetching service config and then decorated with assigning the result to the object.